src/components/Waynet.cpp

```diff
@@ -43,6 +50,274 @@ namespace REGoth
     }
   }
 
+  HWaypoint Waynet::findClosestWaypointTo(const bs::Vector3& position)
+  {
+    if (!hasCachedWaypointPositions())
+    {
+      populateWaypointPositionCache();
+    }
+
+    // No waypoints at all?
+    if (mWaypointPositions.empty())
+    {
+      return {};
+    }
+
+    // Any waypoint would do for initialization, 0 is as good as any other
+    bs::UINT32 nearestIndex = 0;
+    float nearestDistance   = std::numeric_limits<float>().max();
+
+    for (bs::UINT32 i = 0; i < (bs::UINT32)mWaypointPositions.size(); i++)
+    {
+      float thisDistance = (position - mWaypointPositions[i]).squaredLength();
+
+      if (thisDistance < nearestDistance)
+      {
+        nearestDistance = thisDistance;
+        nearestIndex    = i;
+      }
+    }
+
+    return mWaypoints[nearestIndex];
+  }
+
+  HWaypoint Waynet::findSecondClosestWaypointTo(const bs::Vector3& position)
```

Comment:
Since we have to iterate over all waypoints anyway we could probably merge `findClosestWaypointTo` and `findSecondClosestWaypointTo` into one function. The simplest solution would be creating a list of waypoints sorted by distance but that is O(n) in space. A function that returns the closest and second closest waypoint is probably better. If one gets really fancy one could possibly even use templates to get something like
```cpp
template<unsigned int N>
getNClosestWaypointsTo()
```
Just some ideas.

Reply:
I'm not sold on throwing templates at that problem. But getting both nearest and second nearest in one go is a nice efficient idea!

Reply:
It's probably an overkill as long as we don't need more than the second closest point. You could also try to merge it with the freepoint finding logic. Btw: Does it mention in the code what a freepoint is exactly (haven't looked in the wiki yet)?